rules/private/phases/api.bzl

```diff
@@ -5,44 +5,56 @@ load(
 )
 
 def run_phases(ctx, phases):
-    phase_providers = [
-        p[_ScalaRulePhase]
-        for p in [ctx.attr.scala] + ctx.attr.plugins + ctx.attr._phase_providers
-        if _ScalaRulePhase in p
+    toolchain = ctx.toolchains["//rules/scala:toolchain_type"]
+    phase_providers = [toolchain.scala_rule_phases] + [
+        phase_provider[_ScalaRulePhase]
+        for phase_provider in ctx.attr.plugins + ctx.attr._phase_providers
+        if _ScalaRulePhase in phase_provider
     ]
 
     if phase_providers != []:
-        phases = adjust_phases(phases, [p for pp in phase_providers for p in pp.phases])
+        phases = adjust_phases(
+            phases,
+            [phase for phase_provider in phase_providers for phase in phase_provider.phases],
+        )
 
-    gd = {
```

Comment:
Thanks for giving these more descriptive names

Reply:
No problem! Single letter or heavily abbreviated variable names tend to crawl under my skin (with the exception of `i`, `j`, `k`, etc.).